core/commonMain/src/kotlinx/serialization/descriptors/SerialDescriptors.kt

```diff
@@ -88,6 +88,41 @@ public fun PrimitiveSerialDescriptor(serialName: String, kind: PrimitiveKind): S
     return PrimitiveDescriptorSafe(serialName, kind)
 }
 
+/**
+ * Factory to create a new descriptor that is identical to [original] except that the name is equal to [serialName].
+ * Should be used when you want to serialize a type as another non-primitive type.
+ * Don't use this if you want to serialize a type as a primitive value, use [PrimitiveSerialDescriptor] instead.
+ * 
+ * Example:
+ * ```
+ * @Serializable(CustomSerializer::class)
+ * class CustomType(val a: Int, val b: Int, val c: Int)
+ *
+ * class CustomSerializer: KSerializer<CustomType> {
+ *     override val descriptor = SerialDescriptor("CustomType", IntArraySerializer().descriptor)
+ *
+ *     override fun serialize(encoder: Encoder, value: CustomType) {
+ *         encoder.encodeSerializableValue(IntArraySerializer(), intArrayOf(value.a, value.b, value.c))
+ *     }
+ *
+ *     override fun deserialize(decoder: Decoder): CustomType {
+ *         val array = decoder.decodeSerializableValue(IntArraySerializer())
+ *         return CustomType(array[0], array[1], array[2])
+ *     }
+ * }
+ * ```
+ */
+@ExperimentalSerializationApi
+public fun SerialDescriptor(serialName: String, original: SerialDescriptor): SerialDescriptor {
+    require(serialName.isNotBlank()) { "Blank serial names are prohibited" }
+    require(serialName != original.serialName) { "The name of the wrapped descriptor cannot be the same as the name of the original descriptor" }
```

Comment:
I think, it would be a good idea to just return the `original` descriptor instead of throwing an exception.

Reply:
Can't agree here — if one wants keep original name and descriptor, they simply use it without creating wrapper. Name that matches original is clearly a programming mistake here, so we must indicate it

Reply:
I thought more of a dynamic serial name. But yeah, I'm not sure if this is really a case here.